ScoreSaber/UI/Elements/Profile/ProfileDetailView.cs

```diff
@@ -172,6 +295,91 @@ internal async Task ShowProfile(string playerId) {
+                if(_beatmapDownloader == null) {
+                    Plugin.Log.Error("BeatmapDownloader is null, install a mod that injects IScoreSaberBeatmapDownloader");
+                    return;
+                }
+                if (isDownloadingMap) {
+                    return;
+                }
+                isDownloadingMap = true;
+                _panelView.SetPromptInfo("Downloading map...", true);
+                await _beatmapDownloader.DownloadBeatmapAsync(hash, new Action(() => { UnityMainThreadTaskScheduler.Factory.StartNew(async() => 
+                {
+                    _panelView.SetPromptSuccess("Downloaded map!", false, 5f);
+                    _ = UnityMainThreadTaskScheduler.Factory.StartNew(() => SongCore.Loader.Instance.RefreshSongs(true));
+                    while (SongCore.Loader.AreSongsLoading) {
+                        await Task.Delay(200);
+                    }
+                    await Task.Delay(2000);
+                    _ = UnityMainThreadTaskScheduler.Factory.StartNew(() => OpenMap(hash, diff));
+                }); }));
+            }
+        }
+
+        private void OpenMap(string hash, int diff) {
+            Plugin.Log.Info($"Opening map {hash} with diff {diff}");
+            hash = "custom_level_" + hash;
+            var level = SongCore.Loader.BeatmapLevelsModelSO.GetBeatmapLevel(hash);
+            if (level == null) {
+                Plugin.Log.Error($"Level {hash} not found");
+                return;
+            }
+
+            diff = (diff - 1) / 2;
+            BeatmapKey key = level.GetBeatmapKeys().FirstOrDefault(x => diff == (int)x.difficulty);
+            if (key == null) {
+                Plugin.Log.Error($"Difficulty {diff} not found for level {hash}");
+                return;
+            }
+            LevelSelectionFlowCoordinator.State state = new LevelSelectionFlowCoordinator.State(SelectLevelCategoryViewController.LevelCategory.All,
+                                                                                                SongCore.Loader.CustomLevelsPack,
+                                                                                                in key,
+                                                                                                level);
+            //Plugin.Log.Notice($"Opening level {hash} with base game diff {diff}");
+            
+            
+            _soloFreePlayFlowCoordinator.levelSelectionNavigationController._levelCollectionNavigationController._levelCollectionViewController._levelCollectionTableView.SetData(new List<BeatmapLevel>() { level }, _soloFreePlayFlowCoordinator.levelSelectionNavigationController._levelCollectionNavigationController._levelCollectionViewController._levelCollectionTableView._favoriteLevelIds, false, false);
+            _soloFreePlayFlowCoordinator.levelSelectionNavigationController._levelCollectionNavigationController._levelCollectionViewController.SelectLevel(level);
```

Comment:
this really should use some local variables instead of having this extremely long line multiple times.